### PR TITLE
ICU-20578 Guard C++ public headers with C_SHOW_CPLUSPLUS_API, hdrtest cleanup

### PR DESCRIPTION
**Pull request by alonbl**
Continue from #613.

A major cleanup without a functionality change of C public headers.

I left it as a patch series to ease review.

##### ICU-20578 headers: guard all C++ headers using U_SHOW_CPLUSPLUS_API

support including all files in C

cleanup and simplify the `test/hdrtest`

##### ICU-20578 headers: align prologue

remove C++ comments from global blocks

set consistent ordering:
1. copyright
2. changelog
3. `#ifdef` once
4. `\file` documentation block
5. `#includes`

##### ICU-20578 headers: have consistent macro comments

minor format change of comments to reduce changes in next patchset.

##### ICU-20578 headers: make once macro consistent

align all to `__U_HEADER_H__` as unique and consistent.

##### ICU-20578 headers: remove trailing spaces

---

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20578
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added [N/A]

### CODE REVIEW

**Comment by alonbl** [comment]
@jefgen as promised bit love for the public headers, please review this long cleanup patch. I will appreciate quick review as these kind of patches tend to be very hard to rebase.

**Comment by sffc** [comment]
FYI, I think the "new style" for copyright headers has been to use `//` comments. The only reason we retain the old copyright from before 2016 is for historical record.

**Comment by jefgen** [comment]
> ##  file named 'cxxfiles.txt' in this directory.

It might be worth having a comment here that says something like:
If a header fails the C compile test it is likely because the header is a C++ header and isn't properly guarded by the  U_SHOW_CPLUSPLUS_API macro.

---
Refers to: icu4c/source/test/hdrtst/Makefile.in:22 in 8b322eb. [](commit_id = 8b322eb63375159527048f5d0af050d9b01b6d7c, deletion_comment = True)

File: icu4c/source/io/unicode/ustdio.h [hunk 1]
```diff
@@ -21,74 +21,8 @@
...
-#include "unicode/utypes.h"
-#include "unicode/ucnv.h"
-#include "unicode/utrans.h"
-#include "unicode/localpointer.h"
-#include "unicode/unum.h"
-
-#if !UCONFIG_NO_CONVERSION
-
-/*
-    TODO
```
Comment (jefgen) [inline icu4c/source/io/unicode/ustdio.h]: I think fixing/normalizing the include guards (`__U_USTDIO_H__`) is a good idea.
However, I'm less sure about changes like this one, which shuffle around large comment blocks as it tends to "break" the first result from `git blame`.

**Comment by jefgen** [comment]
Thank you @alonbl for the follow-up ticket and for the pull-request!

> I will appreciate quick review as these kind of patches tend to be very hard to rebase.

I've looked over the changes and they seem fine to me.
However, given the amount of churn in this PR it will likely take some time for full review, as I would also need to get opinions and review from others as well.

I think it might potentially be easier if the changes were split up into separate PRs. (You could reuse the ticket number ICU-20578 and have multiple PRs). Though this would be more work for you.

**Comment by alonbl** [comment]
> FYI, I think the "new style" for copyright headers has been to use `//` comments. The only reason we retain the old copyright from before 2016 is for historical record.

I agree that the source of changelog should be in the git log and these can be removed, however, I did not want to make any change... if you give me your OK, I will have another patch that removes the changelog from files.

File: icu4c/source/io/unicode/ustdio.h (hunk 1 above)
Comment (alonbl) [reply icu4c/source/io/unicode/ustdio.h]: I believe that once the file structure is consistent throughout all files, it is easier to manage patches on top. Yes, after files lose consistency re-organization/re-factoring is required, which makes git blame unhappy, however, reaching consistency will allow easier maintenance.

**Comment by alonbl** [comment]
> I think it might potentially be easier if the changes were split up into separate PRs. (You could reuse the ticket number ICU-20578 and have multiple PRs). Though this would be more work for you.

If you look at the branch you will see the branch contains multiple commits, each handles one concern, the problem with github is that it cannot manage review for patch series, this is why I prefer gerrit. It is very difficult to split this into several pull requests as every pull request depend on other pull request and github does not have a mechanism to manage this as well.

In practice, github is not the idle tool for review.

Please tell me what you prefer, we can review patch by patch, meaning I can push the first patch in series, we can review merge and then I will push the next patch in series.

**Comment by alonbl** [comment]
> It might be worth having a comment here that says something like:
> "If a header fails the C compile test it is likely because the header is a C++ header and isn't properly guarded by the U_SHOW_CPLUSPLUS_API macro."

Done

**Comment by jefgen** [comment]
Based on discussion with other ICU-TC members, the only accepted changes to the public headers at this time would be for guarding C++ using the macro `C_SHOW_CPLUSPLUS_API`, the changes to the `hdrtst`, and the trailing white-space fixes.

The opinion of other TC members was that the additional clean-up changes in this cause far too much churn and shouldn't be done at this time.

**Comment by alonbl** [comment]
Thanks! What about of removing the C++ comment style for global scope so that it will be compatible with C?

**Comment by jefgen** [comment]
Thank you @alonbl!

> What about of removing the C++ comment style for global scope so that it will be compatible with C?

Hmm... good question. It seems that the ICU Coding Guidelines permit using them:

http://userguide.icu-project.org/dev/codingguidelines#TOC-C-Coding-Guidelines1

> // Comments
> C++ style // comments may be used in plain C files and in headers that will be included in C files.

The `//` style comments are permitted in C source files as of C99. However, ICU technically doesn't require C99 as a minimum, so this seems a bit odd.
Perhaps it would be worth filing a separate bug for this?

File: icu4c/source/test/hdrtst/Makefile.in [hunk 2]
```diff
@@ -33,135 +32,84 @@ top_builddir = ../..
...
-	  echo '#include "'unicode/$$incfile'.h"' > ht_$$incfile.cpp ; \
-	  echo 'void junk(){}' >> ht_$$incfile.cpp ; \
-	  echo > ht_dep.junk ; \
-          $(COMPILE.cc) -c $(cppflags) ht_$$incfile.cpp 2>&1 | sed -e 's/^.*#error[^"]*"//' | fgrep -v ht_ | tee ht_dep.junk | fgrep -v "$$incfile.h header is obsolete" ; \
-	  RES=$$? ; \
-	  echo $$RES ; \
-	  fgrep -q "$$incfile.h header is obsolete" ht_dep.junk || echo "** FAIL Header unicode/$$incfile.h is not obsoleted properly" ; \
-	  fgrep -q "$$incfile.h header is obsolete" ht_dep.junk || FAIL=1 ; \
-	  rm -f unicode/$$stub ; \
+headertest:
```
Comment (jefgen) [inline icu4c/source/test/hdrtst/Makefile.in]: Thank you for cleaning this up!

**Comment by alonbl** [comment]
> The `//` style comments are permitted in C source files as of C99. However, ICU technically doesn't require C99 as a minimum, so this seems a bit odd.
> Perhaps it would be worth filing a separate bug for this?

Feel free to do so, I tried my best to have a clean patchset that cleanup all the inconsistencies, it is not just the banner, but comments on the `#endif` as well.

I will not go over all headers again,

**Pull request merged**